src/compute-client/src/controller/instance.rs

```diff
@@ -245,18 +245,24 @@ where
             BTreeSet::new()
         };
 
-        // Initialize frontier tracking the new replica.
+        // Initialize frontier tracking the new replica
+        // and clean up any dropped collections that we can
         let mut updates = Vec::new();
+        let mut dropped_collection_ids = Vec::new();
         for (compute_id, collection) in &mut self.compute.collections {
             // Skip log collections not maintained by this replica.
             if collection.log_collection && !maintained_logs.contains(compute_id) {
                 continue;
             }
 
             let read_frontier = collection.read_frontier();
+            if read_frontier.is_empty() {
+                dropped_collection_ids.push(compute_id.clone());
+            }
             updates.push((*compute_id, read_frontier.to_owned()));
         }
         self.update_write_frontiers(id, &updates).await?;
+        self.update_dropped_collections(dropped_collection_ids)?;
```

Comment:
Why do we need to do the check for dropped collections here? Can adding a replica ever lead to a collection fulfilling the drop conditions when it didn't fulfill them before?

Reply:
I am concerned here about a possible interleaving of operations with `rehydrate_replica`. The latter function makes two asynchronous calls: https://github.com/vmarcos/materialize/blob/0313f924e0920e63a382fcc538460aa14b670a5a/src/compute-client/src/controller/instance.rs#L344-L345.

A call to, e.g., `set_read_policy`, could in principle interleave between `remove_replica_state` and `add_replica` setting the collection's read frontier to empty and adding a corresponding `AllowCompaction` command. In this case, we might see no change to the read capabilities in `update_write_frontiers` and the replica being added would never see the AllowCompaction command since it might be eliminated by the subsequent reduction of the command history: https://github.com/vmarcos/materialize/blob/0313f924e0920e63a382fcc538460aa14b670a5a/src/compute-client/src/controller/instance.rs#L277-L278

 I guess we could also simply drop all replicas, then drop a collection, then add a replica to create a similar effect.

Reply:
> A call to, e.g., `set_read_policy`, could in principle interleave between `remove_replica_state` and `add_replica`

I don't think that's possible! `rehydrate_replica` takes a `&mut self`. So according to Rust's borrowing rules, no other mutable borrow of that `ActiveInstance` can be taken until the method call returns. And `ActiceInstance` in turn mutably borrows the underlying `Instance` state. Which means until `rehydrate_replica` returns nothing can mutate the instance state, which setting a read policy would have to do.

Reply:
> I guess we could also simply drop all replicas, then drop a collection, then add a replica to create a similar effect.

In this case we should see the read frontier advance to the empty frontier in `update_read_capabilities` and drop the collection there, no?

Reply:
Thanks for the clarification regarding Rust's borrowing rules! Those would save us from the race in `rehydrate_replica` and I agree with you that `update_read_capabilities` would catch the similar manual interleaving by dropping all replicas.